src/IsolatedExternalsPlugin.ts

```diff
@@ -0,0 +1,195 @@
+  return new ConcatSource(source, loadExternals.toString());
+}
+
+function callLoadExternals(
+  source: Source | string,
+  externals: IsolatedExternalsElement,
+  appName: string
+): ConcatSource {
+  const externalsObj = JSON.stringify(externals);
+  const loadCall = `loadExternals(${externalsObj},`;
+  const appCallback = `function (context) { ${appName}(context); }`;
+  const closeCall = `);`;
+  return new ConcatSource(source, loadCall, appCallback, closeCall);
+}
+
+function selfInvoke(source: Source | string): ConcatSource {
+  return new ConcatSource(`(function() {`, source, `})()`);
+}
+
+function getConfigExternals(
+  externals: ExternalsElement | ExternalsElement[]
+): ExternalsObjectElement {
+  const externalsArray = Array.isArray(externals) ? externals : [externals];
+  const externalsObjects = externalsArray.filter(
+    element => typeof element === 'object' && !(element instanceof RegExp)
+  );
+  const finalExternals = externalsObjects.map(external => {
+    const entries = Object.entries(external);
+    return entries
+      .filter(([, value]) => typeof value === 'string')
+      .reduce<ExternalsObjectElement>(
+        (final: ExternalsObjectElement, [key, value]: [string, string]) => ({
+          ...final,
+          [key]: value
+        }),
+        {}
+      );
+  });
+  const finalElement = finalExternals.reduce<ExternalsObjectElement>(
+    (finalObj: ExternalsObjectElement, obj: ExternalsObjectElement) => ({
+      ...finalObj,
+      ...obj
+    }),
+    {}
+  );
+  return finalElement;
+}
+
+function getExternals(
+  pluginConfig: IsolatedExternalsConfig,
+  compilerExternals: ExternalsObjectElement
+): IsolatedExternals {
+  const externals = Object.entries(pluginConfig).reduce<IsolatedExternals>(
+    (finalExternals: IsolatedExternals, [entryName, pluginExternal]) => {
+      const externalContent = Object.entries(pluginExternal).reduce<
+        IsolatedExternalsElement
+      >(
+        (
+          finalItems: IsolatedExternalsElement,
+          [externalName, externalConfig]: [string, { url: string }]
+        ) => ({
+          ...finalItems,
+          [externalName]: {
+            name: compilerExternals[externalName] as string,
+            ...externalConfig
+          }
+        }),
+        {} as IsolatedExternalsElement
+      );
+      return {
+        ...finalExternals,
+        [entryName]: externalContent
+      };
+    },
+    {} as IsolatedExternals
+  );
+  return externals;
+}
+
+function getTargetAssets(
+  comp: Compilation,
+  config: IsolatedExternalsConfig,
+  externals: IsolatedExternals
+): [string, Source | string][] {
+  const entrypoints = Object.keys(externals)
+    .filter(key => comp.entrypoints.has(key))
+    .map<Entrypoint>(key => comp.entrypoints.get(key) as Entrypoint);
+  const assets = Object.entries<Source | string>(comp.assets);
+  const targetAssets = assets.filter(([name]) =>
+    entrypoints.some(point => point.runtimeChunk.files.includes(name))
+  );
```

Comment:
I could just use `Object.keys` here to make this bit simpler.

Reply:
Turns out this won't work because we're returning the assets as key/value tuples, so we need the keys and the values.